tower-http/src/set_header/request.rs

```diff
@@ -0,0 +1,255 @@
+//! Set a header on the request.
+//!
+//! The header value to be set may be provided as a fixed value when the
+//! middleware is constructed, or determined dynamically based on the request
+//! by a closure. See the [`MakeHeaderValue`] trait for details.
+//!
+//! # Example
+//!
+//! Setting a header from a fixed value provided when the middleware is constructed:
+//!
+//! ```
+//! use http::{Request, Response, header::{self, HeaderValue}};
+//! use tower::{Service, ServiceExt, ServiceBuilder};
+//! use tower_http::set_header::SetRequestHeaderLayer;
+//! use hyper::Body;
+//!
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! # let render_html = tower::service_fn(|request: Request<Body>| async move {
+//! #     Ok::<_, std::convert::Infallible>(Response::new(request.into_body()))
+//! # });
+//! #
+//! let mut svc = ServiceBuilder::new()
+//!     .layer(
+//!         // Layer that sets `User-Agent: my very cool app` on requests.
+//!         //
+//!         // We have to add `::<_, Body>` since Rust cannot infer the body type when
+//!         // we don't use a closure to produce the header value.
+//!         //
+//!         // `if_not_present` will only insert the header if it does not already
+//!         // have a value.
+//!         SetRequestHeaderLayer::<_, Body>::if_not_present(
+//!             header::USER_AGENT,
+//!             HeaderValue::from_static("my very cool app"),
+//!         )
+//!     )
+//!     .service(render_html);
```

Comment:
I believe this example is an HTTP client, but `render_html` sounds rather like a server. Maybe the name should be something like `client` (or `http_client` to be explicit)? Or just passing `hyper::Client::new()` might be fine since the example already uses `hyper` for `hyper::Body`.

Reply:
Good point! I went with `http_client` 937f3a7679058